src/MessagePack.Experimental/CircularReference/CircularReferenceMessagePackSerializerOptions.cs

```diff
@@ -0,0 +1,154 @@
+
+        private int referenceCount;
+
+        private bool isInUse;
+
+        public Span<object> ReferenceSpan => referenceArray.AsSpan(0, referenceCount);
+
+        private CircularReferenceMessagePackSerializerOptions(MessagePackSerializerOptions options)
+            : base(options)
+        {
+            referenceCount = 0;
+            referenceArray = ArrayPool<object>.Shared.Rent(16);
+            isInUse = false;
+        }
+
+        public static CircularReferenceMessagePackSerializerOptions Rent(MessagePackSerializerOptions options)
+        {
+            var element = Table.GetOrCreateValue(options);
+            lock (element)
+            {
+                return element.GetCircularOption(options);
+            }
+        }
+
+        private static readonly ConditionalWeakTable<MessagePackSerializerOptions, WeakTableElement> Table = new ConditionalWeakTable<MessagePackSerializerOptions, WeakTableElement>();
+
+        private sealed class WeakTableElement
+        {
+            public WeakTableElement()
+            {
+                array = ArrayPool<CircularReferenceMessagePackSerializerOptions?>.Shared.Rent(16);
+                Array.Clear(array, 0, array.Length);
+            }
+
+            private CircularReferenceMessagePackSerializerOptions?[] array;
+
+            public CircularReferenceMessagePackSerializerOptions GetCircularOption(MessagePackSerializerOptions options)
+            {
+                for (var i = 0; i < array.Length; i++)
+                {
+                    ref var element = ref array[i];
+                    if (element is null)
+                    {
+                        return element = new CircularReferenceMessagePackSerializerOptions(options) { isInUse = true };
+                    }
+
+                    if (!element.isInUse)
+                    {
+                        element.isInUse = true;
+                        return element;
+                    }
+                }
+
+                var pool = ArrayPool<CircularReferenceMessagePackSerializerOptions?>.Shared;
+                var arrayLength = array.Length;
+
+                var newArray = pool.Rent(arrayLength << 1);
+                Array.Clear(newArray, arrayLength, newArray.Length);
+                Array.Copy(array, newArray, arrayLength);
+                Array.Clear(array, 0, arrayLength);
+                pool.Return(array);
+                array = newArray;
+
+                var answer = array[arrayLength] = new CircularReferenceMessagePackSerializerOptions(options) { isInUse = true };
+                return answer;
+            }
+
+            ~WeakTableElement()
+            {
+                for (var i = 0; i < array.Length; i++)
+                {
+                    var element = array[i];
+                    if (element is null)
+                    {
+                        continue;
+                    }
+
+                    element.Dispose();
```

Comment:
You must *never* dereference an object from a finalizer. Finalizers are *only* for releasing _native_ resources. That's why the Dispose pattern is documented as it is.
Accessing reference types (other than `this`) from a finalizer can crash the process when an AppDomain is being unloaded, since GC/finalization does not guarantee ordering and you could be dereferencing into memory that has already been released.

Reply:
Thank you for your taught!
AppDomain unloading is what I've left behind. Thanks!